infrastructure/cli/commands/helm/rollback.ts

```diff
@@ -0,0 +1,103 @@
+import chalk from 'chalk'
+
+import {helm, releaseName} from './index.js'
+import {history, HistoryRevision} from './history.js'
+import {Environment, namespace} from '../../lib/cluster.js'
+import {makeCommand} from '../../lib/command.js'
+import {confirm} from '../../lib/interactive.js'
+import {Verbosity} from '../../lib/misc.js'
+import {EnvironmentOptions} from '../../lib/options.js'
+
+type ExactRevision = {
+  revision: number
+}
+type DeltaRevision = {
+  steps: number
+}
+type Revision = DeltaRevision | ExactRevision
+
+function coerceRevision(revision: string): Revision {
+  if (revision[0] === '-') {
+    const steps = parseInt(revision.slice(1), 10)
+    if (!isFinite(steps) || steps === 0) {
+      throw new Error('Relative revision must be a whole number of steps')
+    }
+    return {steps}
+  }
+  const revisionNum = parseInt(revision, 10)
+  if (!isFinite(revisionNum)) {
+    throw new Error('Revision must be a whole number')
+  }
+  return {
+    revision: revisionNum,
+  }
+}
+
+export const {command, description, builder} = makeCommand(import.meta.url, 'Roll back to a previous release', {
+  requiredPositionals: [
+    {
+      name: 'revision',
+      description: 'Revision number or number of steps to move (e.g. -2 to go back two revisions)',
+      type: 'string',
+      coerce: coerceRevision,
+    },
+  ],
+})
+
+interface Options extends EnvironmentOptions {
+  revision: Revision
+}
+
+export async function handler({environment, revision, verbosity}: Options): Promise<void> {
+  const releases = await history(environment)
+
+  const currentReleaseIndex = releases.findIndex(({status}) => status === 'deployed')
+  const currentRelease = currentReleaseIndex >= 0 ? releases[currentReleaseIndex] : null
```

Comment:
>const currentRelease = releases.find(release => release.status === 'deployed')
is exactly what i started with, but realised that for the form that rolls back a specified number of steps (as opposed to exact revision) i needed to know the index of the current release in the array.

Reply:
Oh, interesting, I thought `currentReleaseIndex` wasn't used below, how have I missed it 🤦 ?

It's interesting you mention the rollback by steps tho, AFAICT `history()` returns revisions by revision number descending.

Doesn't that mean that the 1st element fo the array is the current release? and therefore if steps is `2` (`-2` to begin with) `releases[2]` would be the release you want to rollback to right?

```
// releases
[0] // current
[1] // previous, -1 steps
[2] // ..., -2 steps
// ...
```

maybe I'm missing something, AFAICT every new release (or rollback) helm creates a new revision with an increased number but I'm probably missing something here.

Reply:
i wasn't sure whether revision numbers **always** go up by 1 so i decided to take the safer route. i could have just done `current revision number - steps to roll back` and assume that's the desired revision number, but wanted to be careful in case somehow a revision could jump around or disappear. in the end, helm history is stored in k8s secrets and one _could_ go in there and mess stuff up.

Reply:
oh I see, they should go up by `1` and pretty much not disappear (unless you delete a release entirely in that case you restart to count from `1`).

Don't worry, I thought I'd mentioned as if may make the code slightly simpler as you would have to check if `steps` is `< releases.length` to avoid going beyond the array boundary (similarly to what you do already) but then you can access target release directly like `releases[steps]` (or something like that)

Anyhow, leave it as it is as the current release index is used below ☺️